[0.6.x] Make it abundantly clear Bunny must run in a fiber

WyriHaximus opened the pull request:
- [X] Add fiber examples for all examples in the readme
- [x] Convert the benchmark to be encapsulated in fibers
- [x] Convert the tutorials to be encapsulated in fibers

@pulmer-cp @mdissington @andrew-demb @menturion @kostirez1 Since you either created or responded/emoted to some message in #192 or #191 I would love your feedback on this PR, addressing that in the documentation and examples. The changes in the readme start here: https://github.com/jakubkulhan/bunny/tree/0.6.x-make-it-abundantly-clear-Bunny-must-run-in-a-fiber?tab=readme-ov-file#always-run-moving-parts-in-a-fiber

WyriHaximus committed:
[0.6.x] Make it abundantly clear Bunny must run in a fiber

kostirez1 commented:
Thank you so much for extending the documentation!

What I was looking for was a way to upgrade `contributte/rabbitmq`, [that relies on the $client->run()](https://github.com/contributte/rabbitmq/blob/f3906e87e672d2362e7097d3d650e233d888152b/src/Consumer/Consumer.php#L62-L76) synchronous helper method [present in 0.5.x](https://github.com/jakubkulhan/bunny/blob/2714c26466d061e1236dbbe9b892324f8be80d63/src/Bunny/Client.php#L167-L172). If I understand it correctly, helper methods are the goal of #178, making the upgrade path much more straightforward. ❤️

mdissington commented:
I think I saw "Cannot switch fibers in current execution context" when I was trying to shut my service down in a signal handler, I take it in that situation it is running outside of a fiber context?

I think I fixed it by calling my shutdown code in a Loop::futureTick().  I wasn't sure if I was doing things the "correct" way but it looks like I may have guessed correctly? :-)

menturion commented:
> I think I saw "Cannot switch fibers in current execution context" when I was trying to shut my service down in a signal handler, I take it in that situation it is running outside of a fiber context?
> 
> I think I fixed it by calling my shutdown code in a Loop::futureTick(). I wasn't sure if I was doing things the "correct" way but it looks like I may have guessed correctly? :-)

That's also exactly my case, terminating a worker via signal handler:

```
        $signals = [SIGINT, SIGTERM, SIGHUP];

        foreach ($signals as $signal) {
            Loop::addSignal($signal, static function () use ($consumerTag): void {
               self::_disconnect($consumerTag);
            });
        }
```

Within the `_disconnect()` method, I call the `cancel(...)` method for the channel object and `disconnect()` for the client object.

andrew-demb commented on README.md:
````diff
@@ -167,6 +167,32 @@ $bunny = new Client($configuration);
 $bunny->connect();
 ```
 
+### Always run moving parts in a fiber
````
Since this is a necessary part of using the library (code can be broken in case of a non-Fiber context https://github.com/jakubkulhan/bunny/issues/191), should this chapter be a bit higher than it is now?

WyriHaximus replied:
Makes perfect sense, also adding a quickstart example right above it

andrew-demb commented on README.md:
```diff
@@ -167,6 +167,32 @@ $bunny = new Client($configuration);
 
+### Always run moving parts in a fiber
+
+The major different between Bunny v0.5 and v0.6 is that it's been rebuild using [`Fibers`](https://reactphp.org/async/). This means that every method on `Client` and `Channel` must be called inside a fiber. In the examples this will be called using a `Loop::futureTick` call. In your applications this might be some other trigger like a HTTP request. As long as somewhere in the direct call stack th ...
```
IMO I think part of "from 0.5 to 0.6" is necessary for the upgrade migration path, but not a generic chapter in the readme

```suggestion
Since v0.6 Bunny has been rebuilt using [`Fibers`](https://reactphp.org/async/). This means that every method on `Client` and `Channel` must be called inside a fiber. In the examples this will be called using a `Loop::futureTick` call. In your applications this might be some other trigger like a HTTP request. As long as somewhere in the direct call stack this call is inside a fiber you'll be good.
```

WyriHaximus replied:
You're right, it makes more sense in the release notes than here. Thanks for the rewording 👍

andrew-demb commented on README.md:
```diff
@@ -167,6 +167,32 @@ $bunny = new Client($configuration);
 
+### Always run moving parts in a fiber
+
+The major different between Bunny v0.5 and v0.6 is that it's been rebuild using [`Fibers`](https://reactphp.org/async/). This means that every method on `Client` and `Channel` must be called inside a fiber. In the examples this will be called using a `Loop::futureTick` call. In your applications this might be some other trigger like a HTTP request. As long as somewhere in the direct call stack th ...
```
> using [`Fibers`](https://reactphp.org/async/).

Does Bunny Bunny require using ReactPHP's implementation of Fiber context?
If not - I suggest hinting to PHP docs at first, and to hint about the recommended abstraction - ReactPHP's

WyriHaximus replied:
It's using the ReactPHP implementation

andrew-demb replied:
I mean - whether all will work with any other implementation or using pure PHP fibers to wrap call of Bunny's client?

andrew-demb commented on README.md:
```diff
@@ -176,6 +202,34 @@ $channel = $bunny->channel();
+    $bunny->connect(); // No perse needed, as the Client::channel() method also does this, but added for completeness sake
+    $channel = $bunny->channel();
+    $channel->queueDeclare('queue_name'); // Queue name
+    $channel->close(); // No perse needed, as the Client::disconnect() method also does this, but added for completeness sake
```
What does it mean "No perse needed"? I'm not sure about the translation.

mdissington replied:
I spotted this, I think he may have meant "Not needed per se, as the Client..."?

andrew-demb replied:
I suggest rephrasing it then to make it clear :+1:

WyriHaximus replied:
So, the client will open a connection when you try to open a channel if it's not connected or in the process of connecting: https://github.com/jakubkulhan/bunny/blob/5a37f769b5c8a1698742fc2394781b37279fe0f1/src/Client.php#L155-L170

Technically, that means there is no need to call `Client::connect()` directly, making Bunny on the connecting side lazy.

Similarly, when calling `Client::disconnect()`, it will close all open channels https://github.com/jakubkulhan/bunny/blob/5a37f769b5c8a1698742fc2394781b37279fe0f1/src/Client.php#L290-L297 . Thus, you don't have to call `Channel::close()` when disconnecting.

Updated the message a bit, but I can also remove them.

andrew-demb commented on README.md:
```diff
@@ -176,6 +202,34 @@ $channel = $bunny->channel();
+      host:     'HOSTNAME',
+      vhost:    'VHOST',    // The default vhost is /
+      user:     'USERNAME', // The default user is guest
+      password: 'PASSWORD', // The default password is guest
```
```suggestion
      password: 'PASSWORD', // The default password is 'guest'
```

andrew-demb commented on README.md:
```diff
@@ -176,6 +202,34 @@ $channel = $bunny->channel();
+  );
+  
+  $bunny = new Client($configuration);
+  Loop::futureTick(async(static function (): void {
```
Why is it necessary to use `Loop::futureTick`, but not `async`? I'm not sure whether it is a necessary detail to be known by the developer for classic use cases (php-fpm with no full async application)

WyriHaximus replied:
> Why is it necessary to use Loop::futureTick, but not async?

`Loop::futureTick` isn't strictly necessary, `async` is tho. You could do `async(static fn etc etc)()` to start the fiber right away, but using a `Loop::futureTick` schedules it right after the event loop has finished all it's incoming working from the kernel.

> I'm not sure whether it is a necessary detail to be known by the developer for classic use cases (php-fpm with no full async application)

It shouldn't matter to them. Will have a look at adding a section for that tomorrow.

andrew-demb reviewed (commented):
This is great, thank you. I left some suggestions to improve it a bit from my PoV

WyriHaximus commented:
> Thank you so much for extending the documentation!
> 
> What I was looking for was a way to upgrade `contributte/rabbitmq`, [that relies on the $client->run()](https://github.com/contributte/rabbitmq/blob/f3906e87e672d2362e7097d3d650e233d888152b/src/Consumer/Consumer.php#L62-L76) synchronous helper method [present in 0.5.x](https://github.com/jakubkulhan/bunny/blob/2714c26466d061e1236dbbe9b892324f8be80d63/src/Bunny/Client.php#L167-L172). If I understand it correctly, helper methods are the goal of #178, making the upgrade path much more straightforward. ❤️

Correct, and @andrew-demb also made some good points on this topic.

Haven't been able to respond to everything, but pushed some changes based on suggestions.

WyriHaximus commented:
> I think I saw "Cannot switch fibers in current execution context" when I was trying to shut my service down in a signal handler, I take it in that situation it is running outside of a fiber context?
> 
> I think I fixed it by calling my shutdown code in a Loop::futureTick(). I wasn't sure if I was doing things the "correct" way but it looks like I may have guessed correctly? :-)

> > I think I saw "Cannot switch fibers in current execution context" when I was trying to shut my service down in a signal handler, I take it in that situation it is running outside of a fiber context?
> > I think I fixed it by calling my shutdown code in a Loop::futureTick(). I wasn't sure if I was doing things the "correct" way but it looks like I may have guessed correctly? :-)
> 
> That's also exactly my case, terminating a worker via signal handler:
> 
> ```
>         $signals = [SIGINT, SIGTERM, SIGHUP];
> 
>         foreach ($signals as $signal) {
>             Loop::addSignal($signal, static function () use ($consumerTag): void {
>                self::_disconnect($consumerTag);
>             });
>         }
> ```
> 
> Within the `_disconnect()` method, I call the `cancel(...)` method for the channel object and `disconnect()` for the client object.

Just pushed updated quickstart examples that cover this, also added the logic to both consumer examples in the readme.
